How to use databases with 12-Factor charms (#2530)
canonical/charmcraft: docs/howto/manage-web-app-charms/use-a-database.rst

.. _use-db-with-12-factor-charms:

> 

Reviewer: @alithethird We've begun adding [meta descriptions](https://canonical-starflow.readthedocs-hosted.com/how-to/add-a-page-meta-description/) to new pages. Would you feel up to writing one?

Reviewer: You're very correct. Sorry, I forgot that page is private. I'll suggest one. :)

Use a database with your 12-factor app charm
============================================

This guide demonstrates how to integrate a MySQL database into your FastAPI
12-factor app. You will learn how to define object-relational mappings (ORM),
handle schema migrations with Alembic, and manage the database integration using Juju.

Prerequisites
-------------

- A working FastAPI application charmed with 12-factor tooling. This guide will be
  following the :ref:`tutorial up until you deploy the app
  <deploy-the-fastapi-app>`.
- A Juju model
- Basic familiarity with `SQLAlchemy <https://www.sqlalchemy.org/>`_ and
  `Alembic <https://alembic.sqlalchemy.org/en/latest/>`_.

Prepare the environment
-----------------------

MySQL requires a specific Python driver. Ensure your ``requirements.txt`` includes

> a MySQL driver compatible with SQLAlchemy, such as ``pymysql``.

Reviewer: We give pymysql as an example, but we actually use it in the doc itself. Could we make it clear that we're using it for the rest of the how-to?

.. code-block::
    :caption: requirements.txt

    fastapi
    sqlalchemy
    alembic
    PyMySQL[rsa]

.. warning::

  When using ``pymysql``, your connection string (handled later by the charm)

>   generally follows the format ``mysql+pymysql://user:pass@host/db``.

Reviewer: I don't feel that this merits a warning.

If we can make it so pymysql is the connector, then I'd make this part of the main text:

```suggestion
When using pymysql, your connection string (handled later by the charm)
follows the format ``mysql+pymysql://user:pass@host/db``.
```

You also need to configure it to read the database URL from the environment,
adhering to 12-factor principles.

.. code-block:: python
    :caption: alembic/env.py

    import os
    from logging.config import fileConfig
    from sqlalchemy import engine_from_config, pool
    from alembic import context
    # Import your models specifically
    from models import Base

    config = context.config

    # 12-factor app: Read config from environment variable
    # The Juju integration will provide this variable
    db_url = os.getenv("MYSQL_DB_CONNECT_STRING")

    if db_url:
        # Convert mysql:// to mysql+pymysql:// for compatibility
        if db_url.startswith("mysql://"):
            db_url = db_url.replace("mysql://", "mysql+pymysql://", 1)
        config.set_main_option("sqlalchemy.url", db_url)

    target_metadata = Base.metadata

    # ... rest of the file ...

Generate a migration
~~~~~~~~~~~~~~~~~~~~

Create the revision file that instructs the database how to create the ``users`` table.
This command requires us to connect to the database, but since we do not have access
to the database, we will use SQLite instead by setting the enviroment variable to

> a non-existent SQLite instance:

Reviewer: > we will use SQLite

Is "use" appropriate? To less attentive readers this might sound like we're bringing in another DB system.

.. code-block:: bash

    export MYSQL_DB_CONNECT_STRING="sqlite:///./test.db"
    alembic revision --autogenerate -m "create users table"

Set up automated updates
~~~~~~~~~~~~~~~~~~~~~~~~

Let's create  a ``migrate.sh`` file to put our migration commands so that 12-Factor
can automatically run your upgrades. This file will be automatically picked up

> by 12-factor tooling and will be run when a database integration event happens

Reviewer: ```suggestion
can automatically run your upgrades. This file will automatically run when a 
database integration event happens
```

          database.py: app/database.py
          models.py: app/models.py
        stage:
        - app/alembic
        - app/alembic.ini
        - app/database.py
        - app/models.py

Now we can pack the file and upload it to local registry:

.. code-block:: bash

    rockcraft pack

    rockcraft.skopeo copy \
      --insecure-policy \
      --dest-tls-verify=false \
      oci-archive:fastapi-hello-world_0.2_$(dpkg --print-architecture).rock \
      docker://localhost:32000/fastapi-hello-world:0.2

.. note::

  If you are using Canonical K8s instead of MicroK8s you need to use this command
  to upload to local registry:

  .. code-block:: bash

    sudo /snap/k8s/current/bin/ctr --address /run/containerd/containerd.sock
    --namespace k8s.io images import --base-name docker.io/library/fastapi-hello-world
    ./fastapi-hello-world_0.2_amd64.rock

Update ``charmcraft.yaml``
--------------------------

Add the ``mysql`` integration to the ``charmcraft.yaml`` file:

.. code-block:: yaml
    :caption: charmcraft.yaml

    requires:
      mysql:
        interface: mysql_client
        optional: false
        limit: 1

Deploy the app
--------------

> Let's deploy the app with Juju:

Reviewer: ```suggestion
Deploy the app with Juju:
```

.. code-block:: bash

    juju deploy ./charm/fastapi-hello-world_amd64.charm --resource app-image=localhost:32000/fastapi-hello-world:0.2

.. note::

  If you are using Canonical Kubernetes instead of MicroK8s you
  need to use this command to deploy your charm:

  .. code-block:: bash

    juju deploy ./charm/fastapi-hello-world_amd64.charm
    --resource app-image=fastapi-hello-world:0.2

Integrate with MySQL
--------------------

> Deploy a MySQL database and integrate it with your application.

Reviewer: ```suggestion
With everything in place, you're ready to deploy a MySQL database and integrate
it with your application.
```

    from sqlalchemy import engine_from_config, pool
    from alembic import context
    # Import your models specifically
    from models import Base

    config = context.config

    # 12-factor app: Read config from environment variable
    # The Juju integration will provide this variable
    db_url = os.getenv("MYSQL_DB_CONNECT_STRING")

    if db_url:
        # Convert mysql:// to mysql+pymysql:// for compatibility
        if db_url.startswith("mysql://"):
            db_url = db_url.replace("mysql://", "mysql+pymysql://", 1)
        config.set_main_option("sqlalchemy.url", db_url)

    target_metadata = Base.metadata

    # ... rest of the file ...

Generate a migration
~~~~~~~~~~~~~~~~~~~~

Create the revision file that instructs the database how to create the ``users`` table:

This command requires us to connect to the database, but since we do not have access
to the database, we will use SQLite instead by setting the enviroment variable to
a non-existent SQLite instance:

.. code-block:: bash

    export MYSQL_DB_CONNECT_STRING="sqlite:///./test.db"
    alembic revision --autogenerate -m "create users table"

Set up automated updates
~~~~~~~~~~~~~~~~~~~~~~~~

> Let's create  a ``migrate.sh`` file to put our migration commands so that 12-Factor

Reviewer 2: ah okay, got it. Thanks for explaining!

> Let's create  a ``migrate.sh`` file to put our migration commands so that 12-Factor

Reviewer: @alithethird Since 12-factor is recognised as a style more than a product, could we explore an alternative like "the charm"?

            DATABASE_URL = DATABASE_URL.replace("mysql://", "mysql+pymysql://", 1)
    engine = create_engine(DATABASE_URL)
    SessionLocal = sessionmaker(autocommit=False, autoflush=False, bind=engine)

    def get_db():
        db = SessionLocal()
        try:
            yield db
        finally:
            db.close()

Create a simple endpoint in ``app.py`` to test the connection:

.. code-block:: python
    :caption: app.py

    from fastapi import FastAPI, Depends, HTTPException
    from sqlalchemy.orm import Session
    import models
    import database

    app = FastAPI()

    @app.get("/")
    async def root():
        return {"message": "Hello World"}

    @app.post("/users/")
    def create_user(username: str, email: str, db: Session = Depends(database.get_db)):
        user = models.User(username=username, email=email)
        db.add(user)
        db.commit()
        db.refresh(user)
        return user

    @app.get("/users/")
    def read_users(skip: int = 0, limit: int = 100, db: Session = Depends(database.get_db)):
        users = db.query(models.User).offset(skip).limit(limit).all()
        return users

Update the ``rockcraft.yaml``

> -----------------------------

Reviewer 2: ```suggestion
Update ``rockcraft.yaml``
-------------------------
```
Or

```suggestion
Update the project file
-----------------------
```

      oci-archive:fastapi-hello-world_0.2_$(dpkg --print-architecture).rock \
      docker://localhost:32000/fastapi-hello-world:0.2

.. admonition:: Note

  If you are using Canonical K8s instead of MicroK8s you need to use this command
  to upload to local registry:
  .. code-block:: bash

    sudo /snap/k8s/current/bin/ctr --address /run/containerd/containerd.sock
    --namespace k8s.io images import --base-name docker.io/library/fastapi-hello-world
    ./fastapi-hello-world_0.2_amd64.rock

Update the ``charmcraft.yaml``
------------------------------

Add the ``mysql`` integration to the ``charmcraft.yaml`` file:

.. code-block:: yaml
    :caption: charmcraft.yaml

    requires:
      mysql:
        interface: mysql_client
        optional: false
        limit: 1

Deploy the app
--------------

Let's deploy the app into Juju:

.. code-block:: bash

    juju deploy ./charm/fastapi-hello-world_amd64.charm --resource app-image=localhost:32000/fastapi-hello-world:0.2

.. admonition:: Note

  If you are using Canonical Kubernetes instead of MicroK8s you
  need to use this command to deploy your charm:
  .. code-block:: bash

    juju deploy ./charm/fastapi-hello-world_amd64.charm

>     --resource app-image=fastapi-hello-world:0.2

Reviewer 2: ```suggestion
.. note:: 

  If you are using Canonical Kubernetes instead of MicroK8s you
  need to use this command to deploy your charm:
  
  .. code-block:: bash

    juju deploy ./charm/fastapi-hello-world_amd64.charm
    --resource app-image=fastapi-hello-world:0.2
```

.. _use-db-with-12-factor-charms:

.. _meta::
    :description: Integrate a MySQL database into your charmed 12-factor app written with FastAPI.

Use a database with your 12-factor app charm
============================================

This guide demonstrates how to integrate a MySQL database into your FastAPI
12-factor app. You will learn how to define object-relational mappings (ORM),
handle schema migrations with Alembic, and manage the database integration using Juju.

Prerequisites
-------------

- A working FastAPI application charmed with 12-factor tooling. This guide will be
  following the :ref:`tutorial up until you deploy the app
  <write-your-first-kubernetes-charm-for-a-fastapi-app-deploy-the-fastapi-app>`.
- A Juju model
- Basic familiarity with `SQLAlchemy <https://www.sqlalchemy.org/>`_ and
  `Alembic <https://alembic.sqlalchemy.org/en/latest/>`_.

Prepare the environment
-----------------------

MySQL requires a specific Python driver. We will use  pymysql, ensure your

> ``requirements.txt`` includes it.

Reviewer: ```suggestion
For Python projects, MySQL requires the pymysql driver. Add it to
``requirements.txt``.
```

    from alembic import context
    # Import your models specifically
    from models import Base

    config = context.config

    # 12-factor app: Read config from environment variable
    # The Juju integration will provide this variable
    db_url = os.getenv("MYSQL_DB_CONNECT_STRING")

    if db_url:
        # Convert mysql:// to mysql+pymysql:// for compatibility
        if db_url.startswith("mysql://"):
            db_url = db_url.replace("mysql://", "mysql+pymysql://", 1)
        config.set_main_option("sqlalchemy.url", db_url)

    target_metadata = Base.metadata

    # ... rest of the file ...

Generate a migration script
~~~~~~~~~~~~~~~~~~~~~~~~~~~

Create the revision file that instructs the database how to create the ``users`` table.
This command requires us to connect to the database, but since we do not have access
to the database, we will use SQLite as an empty database for only this step by setting
the enviroment variable to a non-existent SQLite instance:

.. code-block:: bash

    export MYSQL_DB_CONNECT_STRING="sqlite:///./test.db"
    alembic revision --autogenerate -m "create users table"

Set up automated updates
~~~~~~~~~~~~~~~~~~~~~~~~

> Create  a ``migrate.sh`` script for our migration commands so that 12-Factor

Reviewer: ```suggestion
Create  a ``migrate.sh`` script for the migration commands so that 12-Factor
```

    # See https://documentation.ubuntu.com/rockcraft/1.16.0/reference/extensions/fastapi-framework
    # For questions or help, visit https://matrix.to/#/#12-factor-charms:ubuntu.com

    name: fastapi-hello-world
    base: ubuntu@24.04 # the base environment for this FastAPI application
    version: '0.2' # just for humans. Semantic versioning is recommended
    summary: A summary of your FastAPI application # 79 char long summary
    description: |
        This is fastapi project's description. You have a paragraph or two to tell the
        most important story about it. Keep it under 100 words though,
        we live in tweetspace and your description wants to look good in the
        container registries out there.
    # the platforms this rock should be built on and run on.
    # you can check your architecture with `dpkg --print-architecture`
    platforms:
        amd64:
        # arm64:
        # ppc64el:
        # s390x:

    extensions:
    - FastAPI-framework

    parts:
        alembic:
            plugin: dump
            source: .
            organize:
            alembic: app/alembic
            alembic.ini: app/alembic.ini
            database.py: app/database.py

>             models.py: app/models.py

Reviewer: ```suggestion
                alembic: app/alembic
                alembic.ini: app/alembic.ini
                database.py: app/database.py
                models.py: app/models.py
```
